tests/test_ui_server.py

```diff
@@ -34,3 +42,101 @@ def test_get_vector_colnames(self):
+        from natcap.invest import ui_server
+        test_client = ui_server.app.test_client()
+        response = test_client.post('/getspec', json='sdr')
+        spec = json.loads(response.get_data(as_text=True))
+        self.assertEqual(
+            list(spec),
+            ['model_name', 'module', 'userguide_html',
+             'args_with_spatial_overlap', 'args'])
+
+    def test_get_invest_validate(self):
+        """UI server: get_invest_validate endpoint."""
+        from natcap.invest import ui_server, carbon
+        test_client = ui_server.app.test_client()
+        args = {
+            'workspace_dir': 'foo'
+        }
+        payload = {
+            'model_module': carbon.ARGS_SPEC['module'],
+            'args': json.dumps(args)
+        }
+        response = test_client.post('/validate', json=payload)
+        results = json.loads(response.get_data(as_text=True))
+        expected = carbon.validate(args)
+        # These differ only because a tuple was transformed to a list during
+        # the json (de)serializing, so do the same with expected data
+        self.assertEqual(results, json.loads(json.dumps(expected)))
+
+    def test_post_datastack_file(self):
+        """UI server: post_datastack_file endpoint."""
+        from natcap.invest import ui_server
+        test_client = ui_server.app.test_client()
+        self.workspace_dir = tempfile.mkdtemp()
+        expected_datastack = {
+            'args': {
+                'workspace_dir': 'foo'
+            },
+            'invest_version': '3.10.0',
+            'model_name': 'natcap.invest.carbon'
+        }
+        filepath = os.path.join(self.workspace_dir, 'datastack.json')
+        with open(filepath, 'w') as file:
+            file.write(json.dumps(expected_datastack))
+        response = test_client.post('/post_datastack_file', json=filepath)
+        response_data = json.loads(response.get_data(as_text=True))
+        self.assertEqual(
+            list(response_data),
+            ['type', 'args', 'module_name', 'model_run_name',
+             'model_human_name', 'invest_version'])
+
+    def test_write_parameter_set_file(self):
+        """UI server: write_parameter_set_file endpoint."""
+        from natcap.invest import ui_server
+        test_client = ui_server.app.test_client()
+        self.workspace_dir = tempfile.mkdtemp()
+        filepath = os.path.join(self.workspace_dir, 'datastack.json')
+        payload = {
+            'parameterSetPath': filepath,
+            'moduleName': 'natcap.invest.carbon',
+            'args': json.dumps({
+                'workspace_dir': 'foo'
+            }),
+            'relativePaths': True,
+        }
+        _ = test_client.post('/write_parameter_set_file', json=payload)
+        with open(filepath, 'r') as file:
+            actual_data = json.loads(file.read())
+        self.assertEqual(
+            list(actual_data),
+            ['args', 'invest_version', 'model_name'])
```

Comment:
like above, maybe should be a set? If we do want to guarantee the order of keys in the file produced by `write_parameter_set_file`, lets put that in the docstring

Reply:
order shouldn't matter; changed to a set.